### PR TITLE
Adding product subsystem apis

### PR DESCRIPTION
**Pull request by jkachel**
#### Pre-Flight checklist

- [X] Testing
  - [X] Code is tested

#### What are the relevant tickets?
#305 

#### What's this PR do?
Adds in the REST API for Product retrieval (and adds the scaffolding for that)

#### How should this be manually tested?
Use the standard DRF API explorer (go to /api/v0/products, etc.) once at least one Product has been created (via dj-admin)
pytest tests should also pass

### CODE REVIEW

File: ecommerce/views/v0/__init__.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+        expired_courses = (
+            Course.objects.annotate(
+                runs=Count("courseruns", filter=~Q(courseruns__in=expired_courseruns))
+            )
+            .filter(runs=0)
+            .values_list("id", flat=True)
+        )
+
+        expired_programs = (
```
Comment (annagav) [inline ecommerce/views/v0/__init__.py:48]: I am not sure about the terminology? I feel like expired could mean many different things, like it is not enrollable anymore, or it has finished, or things like that. In MM we used the word 'past' to refer to end date being in the past.

File: ecommerce/views/v0/__init__.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                Q(programruns__isnull=True)
+                | Q(valid_runs=0)
+                | Q(courses__in=expired_courses)
+            )
+            .values_list("id", flat=True)
+            .distinct()
+        )
+
+        return (
+            Product.objects.exclude(
```
Comment (annagav) [inline ecommerce/views/v0/__init__.py:66]: Are we sure we want to exclude past runs of a product? If this queryset is returned for users enrollment info I feel like we would want all of them, right? I might be missing on some context though.

**Review by annagav — CHANGES_REQUESTED** [review]
I left a couple comments, but I feel maybe they I more for @rhysyngsun

File: ecommerce/views/v0/__init__.py (hunk 2 above)
Comment (jkachel) [reply ecommerce/views/v0/__init__.py:66]: Just to add context - my interpretation from the issue was that these API calls would be used to display currently-purchasable items, Pulling enrollment data or order history or things along those lines would be handled elsewhere.

File: ecommerce/views/v0/__init__.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    limit_query_param = 'l'
+    offset_query_param = 'o'
+    max_limit = 50
+
+class ProductViewSet(ReadOnlyModelViewSet):
+    serializer_class = ProductSerializer
+    pagination_class = ProductsPagination
+
+    def get_queryset(self):
+        now = datetime.datetime.now(tz=pytz.UTC)
```
Comment (rhysyngsun) [inline ecommerce/views/v0/__init__.py:34]: Can be `mitol.common.utils.datetime.now_in_utc()`

File: ecommerce/views/v0/__init__.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+        expired_courseruns = CourseRun.objects.filter(
+            enrollment_end__lt=now
+        ).values_list("id", flat=True)
+
+        expired_courses = (
+            Course.objects.annotate(
+                runs=Count("courseruns", filter=~Q(courseruns__in=expired_courseruns))
```
Comment (rhysyngsun) [inline ecommerce/views/v0/__init__.py:42]: Nit: `num_runs` might be clearer.

File: ecommerce/views/v0/__init__.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    max_limit = 50
+
+class ProductViewSet(ReadOnlyModelViewSet):
+    serializer_class = ProductSerializer
+    pagination_class = ProductsPagination
+
+    def get_queryset(self):
+        now = datetime.datetime.now(tz=pytz.UTC)
+
+        expired_courseruns = CourseRun.objects.filter(
```
Comment (rhysyngsun) [inline ecommerce/views/v0/__init__.py:36]: Nit:
```suggestion
        expired_courserun_ids = CourseRun.objects.filter(
```

File: ecommerce/views/v0/__init__.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            enrollment_end__lt=now
+        ).values_list("id", flat=True)
+
+        expired_courses = (
```
Comment (rhysyngsun) [inline ecommerce/views/v0/__init__.py:40]: Nit:
```suggestion
        expired_course_ids = (
```

File: ecommerce/views/v0/__init__.py (hunk 1 above)
Comment (rhysyngsun) [reply ecommerce/views/v0/__init__.py:48]: `unenrollable_program_ids` is probably what I'd go with

File: ecommerce/views/v0/__init__.py (hunk 2 above)
Comment (rhysyngsun) [reply ecommerce/views/v0/__init__.py:66]: I think this is ok as-is for now. If we want to add additional filtering based on a user's enrollments I think we'd want to do that with some optional filtering, but I don't think that's worth adding until we have a use case for it.

**Review by rhysyngsun — CHANGES_REQUESTED** [review]

**Comment by jkachel** [comment]
c1c06dd2674f4011825462737faa58a10e46ad39 updates the wording in a handful of places for clarity and fixes some logic errors in the product filtering code (which was trying to filter on programs rather than program runs).

**Review by rhysyngsun — APPROVED** [review]
LGTM

**Pull request merged**